### ISSUE TITLE
eager_load and includes.references break the inclusion of addtional attributes via select

### ISSUE THREAD

**Issue by bughit** [rails/rails#21197]
``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '4.2.5'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.comments << Comment.create!
    post.comments << Comment.create!

    join_count = Post.joins('left join comments on comments.post_id = posts.id').group(:id).select('posts.*, count(comments.id) as comment_count').first.try(:comment_count)
    includes_count = Post.includes(:comments).references(:comments).select('posts.*, count(comments.id) as comment_count').first.try(:comment_count)
    eager_load_count  = Post.eager_load(:comments).select('posts.*, count(comments.id) as comment_count').first.try(:comment_count)

    assert_equal 2, join_count
    assert_equal 2, includes_count
    assert_equal 2, eager_load_count
  end
end
```

**Comment by bughit** [comment]
also if you remove .references but the query still chooses to do a left join on its own, then it works, don't have an example of this.

**Comment by leoi11** [comment]
+1

**Comment by bughit** [comment]
present in 4.2.5

**Comment by art-solopov** [comment]
Yeah, still here in 5.1.5.
Is there any way to fish out those selected values?

**Comment by kamipo** [comment]
Fixed by #35210.
